localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py

```diff
@@ -81,101 +104,129 @@ def poll_events(self):
+                    # TODO: handle iterator expired scenario
+                    return
+            except Exception:
+                LOG.debug(
+                    f"Attempt {attempts} failed while processing {self.partner_resource_arn} with events: {events}"
+                )
+                attempts += 1
+                # Retry polling until the record expires at the source
+                if self.stream_parameters.get("MaximumRetryAttempts", -1) == -1:
+                    # TODO: handle iterator expired scenario
+                    return
+
+        # Send failed events to potential DLQ
+        abort_condition = abort_condition or "RetryAttemptsExhausted"
+        context = {
+            "condition": abort_condition,
+            "partnerResourceArn": self.partner_resource_arn,
+        }
+        self.send_events_to_dlq(events, context=context)
+        # Update shard iterator if the execution failed but the events are sent to a DLQ
+        self.shards[shard_id] = get_records_response["NextShardIterator"]
+
+    def get_records(self, shard_iterator: str) -> dict:
+        """Returns a GetRecordsOutput from the GetRecords endpoint of streaming services such as Kinesis or DynamoDB"""
+        try:
             get_records_response = self.source_client.get_records(
-                # TODO: double-check cross-account behavior (region should be fine)
-                # StreamARN=self.source_arn,  # differs for DynamoDB but optional
+                # TODO: add test for cross-account scenario
+                # Differs for Kinesis and DynamoDB but required for cross-account scenario
+                **self.stream_arn_param(),
                 ShardIterator=shard_iterator,
                 Limit=self.stream_parameters["BatchSize"],
             )
-            records = get_records_response["Records"]
-            polled_events = self.transform_into_events(records, shard_id)
-            # Check MaximumRecordAgeInSeconds
-            if maximum_record_age_in_seconds := self.stream_parameters.get(
-                "MaximumRecordAgeInSeconds"
-            ):
-                arrival_timestamp_of_last_event = polled_events[-1]["approximateArrivalTimestamp"]
-                now = get_current_time().timestamp()
-                record_age_in_seconds = now - arrival_timestamp_of_last_event
-                if record_age_in_seconds > maximum_record_age_in_seconds:
-                    abort_condition = "RecordAgeExpired"
-            matching_events = self.filter_events(polled_events)
-            # Don't trigger upon empty events
-            if len(matching_events) == 0:
-                # Update shard iterator upon if no records match the filter
-                self.shards[shard_id] = get_records_response["NextShardIterator"]
-                return
-            events = self.add_source_metadata(matching_events)
+            return get_records_response
+        except self.expired_iterator_exception_type() as e:
```

Comment:
The expired iterator exception types are all imported from our API, but this will be a boto client error. I would think that does not work, did you test this?

Reply:
That's correct and I skipped testing because it requires failure destinations to test this code path (implemented in the follow-up PR).

Fixed types in [8feb829](https://github.com/localstack/localstack/pull/11331/commits/8feb829ba1368d48d2495cbd61cfd99b70c1d9d3), also more readable:

```
except self.source_client.exceptions.ExpiredIteratorException as e:
```